### PR TITLE
Add predict example for C++

### PR DESCRIPTION
**Pull request by skylook**
This is a simple predictor which shows how to use c api for image classfication 
It uses opencv for image reading

### CODE REVIEW

**Comment by pertusa** [comment]
Thanks!, I was working on the same this morning, here is the alternative code just in case that you find it useful: https://github.com/pertusa/mxnet_predict_cc. My results are very similar to yours (although I wasn't cropping the image), but after some tests I've found that there is a problem on both implementations. The output is not even similar to that obtained with the corresponding python "mxnet_predict_example.py" for some images, like the one attached. Could you please check it?

borrame

**Comment by piiswrong** [comment]
Thanks! This helps a lot!
A minor suggestion: could you add your build command at the top?

**Comment by piiswrong** [comment]
@pertusa I looked at your code. It pretty good, especially the make file. After we merge this you are welcome to make improvements to it. Thanks for you interest!

**Comment by skylook** [comment]
I have add a Makefile and README file for help, and move every thing to an independent folder.
Thanks to @pertusa & @piiswrong

**Comment by pertusa** [comment]
Thanks @piiswrong , I'll contribute once merged. I've tested the @skylook code, and I get a similar output than with my version, but still very different from the python results. Classifying the vulture image (from my last message) with the ImageNet1K model, I get only a clear peak at the class 904 (window). However, the python script returns a kite as first option, and then some birds. I'm not sure if it a problem of the Makefile (some missing library?), but I've also tried with static linking (updated Makefile in my repo) and this issue remains. Could you try to classify this image and check that the output is consistent?

**Comment by skylook** [comment]
Yes, I met the same problem as @piiswrong , the predict result of c_api seems not very stable. 
Even most of the time it shows the same result with the same image, but sometimes it shows different. 
I do not know if it is a bug or just my miss unstanding with the c api.

**Comment by pertusa** [comment]
Usually non-deterministic results (different outputs from the same input) are due to linkage issues, typically missing libraries. For instance, when using MKL in C++ it is necessary to add many extra flags and libs. So maybe the Makefile is wrong. I can't check it this week, but I'll try to do it ASAP.

**Comment by zhangchen-qinyinghua** [comment]
When I used Theano, some layers with random noise, such as the dropout layer, will make the result non-deterministic. So make sure it is not the reason of your problem.

**Comment by pertusa** [comment]
Good point @zhangchen-qinyinghua, I'll also check this

File: example/cpp/image-classification/image-classification-predict.cpp [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+            mx_float r = (mx_float)(*data++) - mean;
+            
+            *ptr_image++ = r;
+            *ptr_image++ = g;
+            *ptr_image++ = b;
+        }
+    }
+}
+
+void printOutputResult(std::vector<float> data)
```
Comment (tqchen) [inline example/cpp/image-classification/image-classification-predict.cpp]: Please change to Google C++ style ```PrintOutputResult```

Comment (tqchen) [inline example/cpp/image-classification/image-classification-predict.cpp]: If you do not modify data, try pas ```const std::vector<float>& data```

File: example/cpp/image-classification/image-classification-predict.cpp [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+{
+    float bestRes = 0.0;
+    int bestIdx = 0;
```
Comment (tqchen) [inline example/cpp/image-classification/image-classification-predict.cpp]: best_idx

File: example/cpp/image-classification/image-classification-predict.cpp [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+    
+    const mx_uint input_shape_indptr[2] = {(const mx_uint)0, (const mx_uint)4};
+    // ( trained_width, trained_height, channel, num)
+    const mx_uint input_shape_data[4] = {(const mx_uint) 1, (const mx_uint)channels, (const mx_uint)width, (const mx_uint)height};
+    PredictorHandle out= 0; // alias for void *
+    
+    printf("paramStr.len %d\n", paramData.length);
+    
+    //-- Create Predictor
+    MXPredCreate((const char* ) jsonData.buffer,
```
Comment (tqchen) [inline example/cpp/image-classification/image-classification-predict.cpp]: No need to cast here

File: example/cpp/image-classification/image-classification-predict.cpp [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+                 (const char* ) paramData.buffer,
+                 (size_t) paramData.length,
```
Comment (tqchen) [inline example/cpp/image-classification/image-classification-predict.cpp]: use ```static_cast<size_t>``` instead

File: example/cpp/image-classification/image-classification-predict.cpp [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+    int width = 224;
+    int height = 224;
+    int channels = 3;
+    
+    const mx_uint input_shape_indptr[2] = {(const mx_uint)0, (const mx_uint)4};
+    // ( trained_width, trained_height, channel, num)
+    const mx_uint input_shape_data[4] = {(const mx_uint) 1, (const mx_uint)channels, (const mx_uint)width, (const mx_uint)height};
+    PredictorHandle out= 0; // alias for void *
+    
+    printf("paramStr.len %d\n", paramData.length);
```
Comment (tqchen) [inline example/cpp/image-classification/image-classification-predict.cpp]: Consider use ```LOG(INFO) << param_str.length=``` << param_data.length

**Comment by tqchen** [comment]
Thanks for the good job, can you please make a pass on code style to make it consistent with rest of the project ?  i.e. Google's C style.

Check the  ```lint``` job in main Makefile, which can help auto checks the code style

Thanks!

**Comment by pertusa** [comment]
I've found the problem of the inconsistent output. OpenCV data array is stored as BGR,BGR,BGR, whereas MXPredSetInput expects the array as RRRRRR...,GGGGGGG...,BBBBBBB.... I've updated my code in https://github.com/pertusa/mxnet_predict_cc accordingly and now it works. @tqchen, could you please change your loop in getMeanFile?

About the non-deterministic output, dropout layers give some random values as @zhangchen-qinyinghua commented. This is not a serious issue as the expected results are usually quite similar but some frameworks like Caffe have some layers that can be discarded int the TEST phase (usually dropouts). I wonder if there is something similar in MXNet, but if not it could be a good feature.

**Comment by piiswrong** [comment]
@pertusa Are you sure it's not BBBBBB...,GGGGGGGG....,RRRRRRRR......

**Comment by pertusa** [comment]
Yes, I've checked it printing the data array received in MXPredSetInput() of c_predict_api.cc. With the python call from (mxnet_predict_example.py) and a red image, the first elements in the array are the ones with the highest values, so it's RRR...GGG...BBB...

**Comment by skylook** [comment]
Some small modifications:
1. Updated the code according to Google C Style and @tqchen 's suggestions.
2. Fix the image loading error and add synset result showing as @pertusa .

**Comment by pertusa** [comment]
Good work Skylook, thanks!

File: example/cpp/image-classification/image-classification-predict.cpp [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+     output.push_back(lemma);
+   }
+   
+   fi.close();
+   
+   return output;
+}
+
+void PrintOutputResult(const std::vector<float>& data, const std::vector<std::string>& synset) {
+    if ( data.size() != synset.size() ) {
```
Comment (tqchen) [inline example/cpp/image-classification/image-classification-predict.cpp]: No space before bracket

File: example/cpp/image-classification/image-classification-predict.cpp [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+    
+    resize(im_ori, im, resize_size);
+    
+    // Better to be read from a mean.nb file
+    float mean = 117.0;
+    
+    size = im.rows * im.cols * 3;
+
+    mx_float* ptr_image_r = image_data;
+    mx_float* ptr_image_g = image_data + size/3;
```
Comment (tqchen) [inline example/cpp/image-classification/image-classification-predict.cpp]: space between operands ```size / 3```

File: example/cpp/image-classification/image-classification-predict.cpp [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+    mx_float* ptr_image_b = image_data + size/3*2;
+    
+    for (int i = 0; i < im.rows; i++) {
+        uchar* data = im.ptr<uchar>(i);
+        
+        for (int j = 0; j < im.cols; j++) {
+            mx_float b = (mx_float)(*data++) - mean;
```
Comment (tqchen) [inline example/cpp/image-classification/image-classification-predict.cpp]: Use static_cast<mx_float>()

File: example/cpp/image-classification/image-classification-predict.cpp [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+                 num_input_nodes,
+                 input_keys,
+                 input_shape_indptr,
+                 input_shape_data,
+                 &out
+                 );
+    
+    
+    // Just a big enough memory 1000x1000x3
+    mx_float* image_data = new mx_float[width * height * channels];
```
Comment (tqchen) [inline example/cpp/image-classification/image-classification-predict.cpp]: optional, consider directly use ```std::vector<mx_float>``` for data, for RAII.

**Comment by tqchen** [comment]
Thanks for being patient on this and great contribution. Can you please do a lint on your file? like

```bash
dmlc-core/script/lint.py cpp mxnet cpp example/cpp
```
You can add this path to https://github.com/dmlc/mxnet/blob/master/Makefile#L145 so it is automatically checked by travis This should detect most of common things linter complained on, I commented a few. This is just to make sure the project have consistent programming style :)

After this is done, please 
- Rename the suffix from .cpp to .cc
- update the PR to add yourself to https://github.com/dmlc/mxnet/blob/master/CONTRIBUTORS.md
- Squash the commits to meaningful ones via http://mxnet.readthedocs.org/en/latest/contribute.html#how-to-combine-multiple-commits-into-one 
   - So each commit means meaningful progress.

Then it is good to be merged.

**Comment by tqchen** [comment]
@skylook can you update the code accordingly?

**Comment by caprice-j** [comment]
@skylook Hi, Thanks for improving my code in #531. 
May I add my name after you to the header of image-classification-predict.cpp?

**Comment by skylook** [comment]
I am working on it, sorry for being too busy :)

**Comment by skylook** [comment]
I got some sample code from my workmates and didnot notice your github, I will sure to add your name when I finished :) @caprice-j

**Comment by skylook** [comment]
Hi, I finally have time to do these following changes:

1. Change code style to pass cpplint check.
2. Update README file to add more details of building and usage of the code. 
3. Combine multiple commits into one.

Hopefully this patch is ready to merge :) @tqchen

**Comment by piiswrong** [comment]
Looks good to me. @tqchen

**Comment by tqchen** [comment]
Great Job! Thanks for the contribution and make mxnet better for everyone!

**Comment by caprice-j** [comment]
@skylook Thanks for your kindness.

**Pull request merged**